### PR TITLE
METRON-1405: Add Boyer-Moore majority vote algorithm to Stellar

### PR DESCRIPTION
**Pull request by mmiklavc**
## Contributor Comments

Addresses https://issues.apache.org/jira/browse/METRON-1405

A proper majority requires 2 passes, however you can get a sense of plurality by leveraging a single pass model.

Test plan soon to follow.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [ ] Have you included steps or a guide to how the change may be verified and tested manually?
- [ ] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java [hunk 1]
```diff
@@ -220,4 +219,156 @@ public Object apply(List<Object> args) {
...
+      return counter;
+    }
+  }
+
+  @Stellar(namespace = "BOYERMOORE"
+      , name = "PLURALITY"
+      , description = "Calculates the item with current plurality in a Boyer-Moore list. [Boyer-Moore](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_majority_vote_algorithm)"
+      , params = {
+      "state - state holder for list of values. If null, add will initialize a new state value."
+  }
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: This is confusing.  It is the same description as ADD() but I think that ADD is correct and this should be "Returns the object that has plurality" ?

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java [hunk 2]
```diff
@@ -220,4 +219,156 @@ public Object apply(List<Object> args) {
...
+        }
+      }
+      return m;
+    }
+
+    public Object add(Object item, Long counter) {
+      if (item != null) {
+        if (this.counter == 0) {
+          m = item;
+          this.counter = counter;
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: So, Stellar will parse and create Numbers, I'm wondering if there is a possibility of strange errors or rounding here.  Like there are two not equal numbers but they end up equal through the conversion process.  Does that make sense?

**Comment by ottobackwards** [comment]
This tests fails, I think something is wrong.

```java
 @Test
  public void boyerMoore_calculates_plurality_from_list_of_mixed_objects_o() throws Exception {
    List<Object> items = Arrays.asList(1,1,1,1, "orange", "orange", "jello", "jello", "jello");
    BoyerMooreState state = (BoyerMooreState) new BoyerMooreAdd().apply(Arrays.asList(null, items));
    Assert.assertThat(new BoyerMoorePlurality().apply(ImmutableList.of(state)), CoreMatchers.equalTo(1));
  }
```

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java (hunk 1 above)
Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: Good catch

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java (hunk 2 above)
Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: It's true, I make no effort to massage data types here - if you want to interpret a 1 as a string "1", then they will be considered distinct. I didn't have a suitable mechanism for determining types. I thought it was reasonable for a user to be able to make that type cast as part of their Stellar expression building, but I'm certainly open to suggestions.

Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: AFA rounding errors - the counter you're seeing there are internal state. In the case of merging multiple states together, it will effectively merge the counts to nominate a new plurality leader. If we could do this in 2 passes, you'd have a guaranteed majority, but I thought it was still useful to see what item has the most without necessarily being a majority. I'll add some use case documentation this week as I think this might be useful in the profiler.

**Comment by mmiklavc** [comment]
I'll look into the test as it's passing locally for me in my IDE.

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java (hunk 2 above)
Comment (ottobackwards) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: no, i'm talking about  object.equals(object2)  when the type of the objects is determined by antlr etc

**Comment by ottobackwards** [comment]
Just in case:  Note that that test is modified from your test ( _o )

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java (hunk 2 above)
Comment (mmiklavc) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: I know we do this in a number of places where the Objects provided via the inbound List are handed off directly to a java.util.List without any casting or special parsing, e.g. 
* BloomFilter -> https://github.com/apache/metron/blob/master/metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/DataStructureFunctions.java#L48
* Data structure functions such as List -> https://github.com/apache/metron/blob/master/metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/DataStructureFunctions.java#L189
* HLLP -> https://github.com/apache/metron/blob/master/metron-analytics/metron-statistics/src/main/java/org/apache/metron/statistics/approximation/HyperLogLogPlusFunctions.java#L49

Is that what you're asking?

**Comment by mmiklavc** [comment]
Yes - good catch - this not actually going to work without some modification. The count zeroes out when a majority is held by non-m values and gets reset to a new non-plurality, non-majority object.

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java (hunk 2 above)
Comment (ottobackwards) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/dsl/functions/MathFunctions.java]: No, more of something that is prob. an edge case, such as fields from input that are not normalized, so they come in as 1 or 1.0 or something and will be unequal.

**Comment by ottobackwards** [comment]
@Test
  public void boyerMoore_calculates_plurality_from_list_of_values_o() throws Exception {
    List<Integer> items = Arrays.asList(1, 1, 1,1, 2,2,3,3,4,4,4);
    BoyerMooreState state = null;
    for (Integer item : items) {
      state = (BoyerMooreState) new BoyerMooreAdd().apply(Arrays.asList(state, item));
    }
    Assert.assertThat(new BoyerMoorePlurality().apply(ImmutableList.of(state)), CoreMatchers.equalTo(1));
  }

as well

**Comment by mmiklavc** [comment]
Pulling this PR until I can resolve and underlying issue with the counting semantics, however I believe this may by theoretically impossible without a 2-pass model.
